ompi/mca/coll/ftagree/coll_ftagree_earlyreturning.c

```diff
@@ -0,0 +1,3367 @@
+    int *new_dead;
+    int *ack_failed;
+
+    assert(MCA_BTL_TAG_FT_AGREE == tag);
+    assert(1 == descriptor->des_segment_count);
+    (void)cbdata;
+
+    frag = (era_frag_t*)descriptor->des_segments->seg_addr.pval;
+
+    if( frag->msg_len == frag->frag_len ) {
+        assert(frag->frag_offset == 0);
+        msg_bytes = frag->bytes;
+    } else {
+        src_hash = hash_name(frag->src);
+        opal_mutex_lock(&era_incomplete_msg_mutex);
+        if( opal_hash_table_get_value_uint64(&era_incomplete_messages, src_hash, &value) == OMPI_SUCCESS ) {
+            msg_table = (opal_hash_table_t*)value;
+        } else {
+            msg_table = OBJ_NEW(opal_hash_table_t);
+            opal_hash_table_init(msg_table, 3 /* This should be very small: few messages should fly in parallel */);
+            opal_hash_table_set_value_uint64(&era_incomplete_messages, src_hash, (void*)msg_table);
+        }
+
+        if( opal_hash_table_get_value_uint64(msg_table, frag->msg_seqnum, &value) == OMPI_SUCCESS ) {
+            incomplete_msg = (era_incomplete_msg_t*)value;
+        } else {
+            incomplete_msg = (era_incomplete_msg_t*)malloc(frag->msg_len + sizeof(unsigned int));
+            incomplete_msg->bytes_received = 0;
+            opal_hash_table_set_value_uint64(msg_table, frag->msg_seqnum, (void*)incomplete_msg);
+        }
+        opal_mutex_unlock(&era_incomplete_msg_mutex);
+
+        memcpy( incomplete_msg->bytes + frag->frag_offset,
+                frag->bytes,
+                frag->frag_len );
+        incomplete_msg->bytes_received += frag->frag_len;
+
+        /** We receive the messages in order */
+        if( incomplete_msg->bytes_received == frag->msg_len ) {
+            msg_bytes = incomplete_msg->bytes;
+            opal_mutex_lock(&era_incomplete_msg_mutex);
+            opal_hash_table_remove_value_uint64(msg_table, frag->msg_seqnum);
+            /** We leave msg_table into the global table, as we will receive more messages */
+            opal_mutex_unlock(&era_incomplete_msg_mutex);
+        } else {
+            /** This message is incomplete */
+            return;
+        }
+    }
+
+    msg_header = (era_msg_header_t *)msg_bytes;
+    msg_bytes += sizeof(era_msg_header_t);
+
+    if( ERA_VALUE_BYTES_COUNT(&msg_header->agreement_value_header) > 0 ) {
+        value_bytes = msg_bytes;
+        msg_bytes += ERA_VALUE_BYTES_COUNT(&msg_header->agreement_value_header);
+    } else {
+        value_bytes = NULL;
+    }
+
+    if( msg_header->agreement_value_header.nb_new_dead > 0 ) {
+        new_dead = (int*)msg_bytes;
+        msg_bytes += msg_header->agreement_value_header.nb_new_dead * sizeof(int);
+    } else {
+        new_dead = NULL;
+    }
+
+    if( msg_header->nb_ack > 0 ) {
+        ack_failed = (int*)msg_bytes;
+    } else {
+        ack_failed = NULL;
+    }
+
+#if defined(PROGRESS_FAILURE_PROB)
+#pragma message("Hard coded probability of failure inside the agreement")
```

Comment:
Do we want to keep this feature (injection of failures) inside the PR?

Reply:
It is not unheard off to have some debugging tooling that is disabled and can be turned on only with changing source code (TCP_DUMP, for example). @bosilca what do you think?

Reply:
I don't expect anybody will care about such a tiny issue, and it can be handy in the future to have such a capability built-in.